### ISSUE TITLE
Compatibility with the CSS custom function and mixin proposal

### ISSUE THREAD

**Issue by stof** [sass/sass#3787]
- [x] [Proposal](https://github.com/sass/sass/blob/main/accepted/css-function-mixin-preparation.md) (https://github.com/sass/sass/pull/3812)
- [x] Update official spec (https://github.com/sass/sass/pull/3854)
- **Deprecation**
   - [x] [Dart Sass](https://github.com/sass/dart-sass/issues/2197) (https://github.com/sass/dart-sass/pull/2230)
   - [x] [Tests](https://github.com/sass/sass-spec/issues/1962) (https://github.com/sass/sass-spec/pull/1982)
   - [x] [Website](https://github.com/sass/sass-site/issues/979) (https://github.com/sass/sass-site/pull/1028)
- **CSS Compatibility** (`@function`)
   - [x] Spec (https://github.com/sass/sass/pull/4137)
   - [x] [Dart Sass](https://github.com/sass/dart-sass/issues/2197) (https://github.com/sass/dart-sass/pull/2655)
   - [x] [Tests](https://github.com/sass/sass-spec/issues/1962) (https://github.com/sass/sass-spec/pull/2087)
   - [x] [Website](https://github.com/sass/sass-site/issues/979) (https://github.com/sass/sass-site/pull/1494)
- **CSS Compatibility** (`@mixin`)
   - [ ] Spec
   - [ ] [Dart Sass](https://github.com/sass/dart-sass/issues/2197)
   - [ ] [Tests](https://github.com/sass/sass-spec/issues/1962)
   - [ ] [Website](https://github.com/sass/sass-site/issues/979)

---

@mirisuzanne is working on a proposal for CSS custom functions and mixins at https://css.oddbird.net/sasslike/mixins-functions/

The suggested syntax uses `@function` and `@mixin`, which are the same at-rules than Sass, and so would clash with it right now.

The CSS proposal requires CSS custom functions and mixins to use `dashed-ident`, i.e. an identifier starting with `--`, to avoid clashes with built-in CSS functions (which never start with `--`).
However, Sass currently allows defining function and mixin names starting with `--`, as shown in https://sass-lang.com/playground/#eJwzNHRIK81LLsnMz1PQ1S1JLS5R0NBUqOZSUHAoSi0pLcpTUE9Wt+aq5eJyyM2syIQpMoKpSrFSSAVLJ4K5SVZQBRqa1iAzMvOSc0pTUmG6gCoBJ/UfhA==

Should we deprecate defining Sass members named using `dashed-ident` to reduce the risk of clashes with CSS ?

**Comment by nex3** [comment]
Yep, we should definitely deprecate these and plan to eventually pass anything that starts with `@function <dashed-ident>` or `@mixin <dashed-ident>` through as an unaltered at-rule. Do you want to write up a proposal?

**Comment by stof** [comment]
I can work on it.

High level plan:

- deprecation period where dashed-ident as function or mixin names in Sass files trigger a deprecation warning
- this deprecation period will be announced to end either in Dart Sass 2.0 or once CSS compatibility warrants an earlier BC break (which depends on when the CSS custom function proposal ships in browsers), whatever comes first. After the deprecation period, `@function <dashed-ident>` and `@mixin <dashed-ident>` will be parsed as unaltered at-rule.
- add immediately the support of `@function <dashed-ident>` and `@mixin <dashed-ident>` in CSS files as unaltered at-rule (keeping the parse error for other `@function` or `@mixin` like for other Sass features)

Does this plan looks good ?

**Comment by stof** [comment]
Note that in Sass files, there is already a way to output a CSS custom function today: using an interpolated at-rule: https://sass-lang.com/playground/#eJwzNHRQrlZPK81LLsnMz1OvVdDVLUktLlGo5lJQKEotLs0psVJItOaqBQAEqgyt

**Comment by nex3** [comment]
Yep, looks good!

**Comment by nex3** [comment]
This proposal is now out for review. I'll give it a month for public comment.

**Comment by stof** [comment]
the proposal looks good to me (and sorry for not doing the proposal myself)

**Comment by nex3** [comment]
It's been a month, so I'm going to mark this accepted and start on implementation.

**Comment by stof** [comment]
CSS custom functions have shipped in Chrome 139 (CSS mixins are not implemented yet)

**Comment by nex3** [comment]
There's an interesting decision to make here about how we want to syntactically handle `@function`'s `result` property. The CSS spec allows it to contain any `<declaration-value>`, just like custom properties, and the values it produces are similarly accessible to JS if used in a custom property (as in `--foo: --bar()`). For custom property declarations, we fall back on avoiding SassScript syntax for this reason, and require it to only be injected via `#{}`.

The reasons for this are twofold: first, philosophically, we want SCSS to remain as close to a CSS superset as possible, in the sense that any valid and meaningful CSS stylesheet should also be a valid SCSS stylesheet with the same semantics. Second, more pragmatically, at the time custom properties were first introduced, Polymer.js introduced a polyfill for an early draft of CSS mixins that used the following syntax:

```css
:root {
  --mixin: {
    color: red;
  }
}
```

However, times have changed since then. This Polymer syntax is long-dead and since then there have been essentially zero new uses of the wider flexibility of this syntax of any note with the sole exception of writing a property whose value is a single space which is used for toggles—something that could easily be compatible with plain CSS.

On the other hand, this syntactic difference does cause real confusion and annoyance for users. So I'm wondering: are our CSS compatibility gains worth the hassle? While compatibility is a core design goal, JS API visibility isn't necessarily something we have to be chained to (certainly we already rearrange stylesheets in a way that's visible to the CSSOM). In terms of compatibility _within a stylesheet_ the cases are far narrower—technically `--comma: ,;` is valid and allows users to write `var(--comma)` instead of a literal comma, but this is certainly no more substantial of an incompatibility than the existing interpretation of `#{}` as interpolation.

Do the question at hand is, do we still consider the special custom property syntax to be valuable? If so, we probably want to adopt it for `result:` in `@function` as well. But if not, we should start considering a path to move (back) to normal SassScript syntax in other custom property values.

**Comment by mirisuzanne** [comment]
I don't know of any common uses that require the special syntax handling, but I imagine there may be some uncommon ones - such as swapping out operators in a `calc()` perhaps? It might be worth doing some broader survey, or reaching out to some CSS artists. 

I don't know if there's a way to make normal SassScript behavior the default, while still providing some path for the extended syntax when desired (does `unquote($string)` get you there)? My primary concern with the existing variable rules would be if breaking backwards-compat is more pain than it's worth – even if the feature isn't being used extensively.

**Comment by nex3** [comment]
`unquote()` will work as an escape hatch, as would interoplation with a quoted string. If we did decide to move off of dedicated parsing for custom properties, we'd probably migrate users from `--operator: +` to `--operator: #{"+"}`.

I've posted [this poll](https://front-end.social/@sass/115272830411739760) to gather data about how people are using the feature currently. If folks could boost it or forward to people who are likely to be using these things heavily or weirdly, I'd appreciate it. I'll probably hold off on making any decisions until it's done, although for the time being I'm leaning towards parsing `result:` as SassScript.

**Comment by nex3** [comment]
The poll has closed, and about one third of respondents who report using CSS variables at all report using non-SassScript syntax for them. This is surprising to me, and unfortunately no one replied with examples of _what_ syntax they're using specifically.  This does suggest to me that it's probably not a good idea to move away from our specialized custom property syntax; it's unclear to me whether that means we should make `result:` operate like a custom property or not because I don't have a clear understanding of the use-cases and whether they're relevant to functions.

**Comment by mirisuzanne** [comment]
Yeah, that's rough. I'm also curious if people _know that well_ what would be valid CSS syntax. Would be very helpful to have batter data.

**Comment by nex3** [comment]
We got one more specific answer: a [user reported](https://indieweb.social/@djenz/115332127961066281) using `--dynamic-details_disabled: (min-width: #{$dynamic-details-breakpoint}px)` to "[pass] a breakpoint to JS, so that the breakpoint is not defined in multiple places", presumably via [`matchMedia()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia). I think this is a reasonably compelling use-case (certainly moreso than the polymer syntax): at-rule syntax that's native to CSS but usable in JS. That's enough to be a convincing case not to change the current custom property syntax.

The question then is, how likely are people to want to write that from a custom function? It's definitely _conceivable_ that someone would write `@function --my-query(--breakpoint) {result: (max-width: var(--breakpoint)}}` (likely with a much more complex query), so perhaps the principle of conservatism means that we should in the end preserve the ability to write CSS syntax.

Another factor here is that, because CSS functions are already a form of abstraction, they may be less likely than the average CSS feature to be something users want to inject Sass variables into. This would also speak in favor of supporting the more general CSS syntax.

**Comment by nex3** [comment]
Thinking about this more, I do think it's best to treat `result` the same as CSS custom properties. This means we can move forward with the draft PRs as-is.
